src/query/runner/ipa.rs

```diff
@@ -1,121 +1,89 @@
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + Serializable,
+    IPAInputRow<F, MatchKey, BreakdownKey>: Serializable;
+
+impl<F, C, S> IpaQuery<F, C, S>
+where
+    F: PrimeField,
+    AdditiveShare<F>: Serializable,
+    C: UpgradableContext,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + Serializable,
+    IPAInputRow<F, MatchKey, BreakdownKey>: Serializable,
+{
+    pub fn new(config: IpaQueryConfig) -> Self {
+        Self(config, PhantomData)
     }
+}
 
-    pub async fn malicious_run(
-        &self,
-        ctx: MaliciousContext<'_>,
-        field: FieldType,
+impl<F, C, S, SB> IpaQuery<F, C, S>
+where
+    C: UpgradableContext + Send,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S> + RandomBits<F, Share = S>,
+    S: LinearSecretSharing<F>
+        + BasicProtocols<C::UpgradedContext<F>, F>
+        + Reshare<C::UpgradedContext<F>, RecordId>
+        + Serializable
+        + DowngradeMalicious<Target = AdditiveShare<F>>
+        + 'static,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = SB>,
+    SB: LinearSecretSharing<Gf2>
+        + BasicProtocols<C::UpgradedContext<Gf2>, Gf2>
+        + DowngradeMalicious<Target = AdditiveShare<Gf2>>
+        + 'static,
+    F: PrimeField,
+    IPAInputRow<F, MatchKey, BreakdownKey>: Serializable,
+    ShuffledPermutationWrapper<S, C::UpgradedContext<F>>: DowngradeMalicious<Target = Vec<u32>>,
+    MCCappedCreditsWithAggregationBit<F, S>:
+        DowngradeMalicious<Target = MCCappedCreditsWithAggregationBit<F, AdditiveShare<F>>>,
+    MCAggregateCreditOutputRow<F, S, BreakdownKey>:
+        DowngradeMalicious<Target = MCAggregateCreditOutputRow<F, AdditiveShare<F>, BreakdownKey>>,
+    AdditiveShare<F>: Serializable,
+{
+    pub async fn execute<'a>(
+        self,
+        ctx: C,
         input: ByteArrStream,
-    ) -> Box<dyn ProtocolResult> {
-        match field {
-            #[cfg(any(test, feature = "weak-field"))]
-            FieldType::Fp31 => Box::new(
-                self.malicious_run_internal::<crate::ff::Fp31, MatchKey, BreakdownKey>(ctx, input)
-                    .await
-                    .expect("IPA query failed"),
-            ),
-            FieldType::Fp32BitPrime => Box::new(
-                self.malicious_run_internal::<Fp32BitPrime, MatchKey, BreakdownKey>(ctx, input)
-                    .await
-                    .expect("IPA query failed"),
-            ),
+    ) -> Result<Vec<MCAggregateCreditOutputRow<F, AdditiveShare<F>, BreakdownKey>>, Error> {
+        let Self(config, _) = self;
+
+        let mut input =
+            input.align(<IPAInputRow<F, MatchKey, BreakdownKey> as Serializable>::Size::USIZE);
+        let mut input_vec = Vec::new();
+        while let Some(data) = input.next().await {
+            input_vec.extend(IPAInputRow::<F, MatchKey, BreakdownKey>::from_byte_slice(
+                &data.unwrap(),
+            ));
```

Comment:
The big win is that there's only one copy of this code that does `input.align` and the subsequent processing of the inputs. It's only a few lines right now, but it's about to get a bit more involved.